File: matrix_mult.cpp
#include <iostream>
#include <random>
#include <chrono>
#include <mkl.h>
#include <cstdlib>

void initialize_matrix(double* matrix, int size) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<double> dis(-1.0, 1.0);

    for (int i = 0; i < size * size; ++i) {
        matrix[i] = dis(gen);
    }
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <matrix_size>" << std::endl;
        return 1;
    }

    int n = std::atoi(argv[1]);
    if (n <= 0) {
        std::cerr << "Matrix size must be positive" << std::endl;
        return 1;
    }

    MKLVersion version;
    mkl_get_version(&version);

    std::cout << "=== Intel MKL Version Information ===" << std::endl;
    std::cout << "MKL version: " << version.MajorVersion << "." << version.MinorVersion << std::endl;

    // Allocate memory for matrices
    double* A = new double[n * n];
    double* B = new double[n * n];
    double* C = new double[n * n];

    // Initialize matrices with random values
    initialize_matrix(A, n);
    initialize_matrix(B, n);

    // Set alpha and beta for GEMM operation (C = alpha*A*B + beta*C)
    double alpha = 1.0;
    double beta = 0.0;

    // Measure time for matrix multiplication
    auto start = std::chrono::high_resolution_clock::now();

    // Perform matrix multiplication: C = A * B
    // Note: MKL uses column-major order, so we compute B^T * A^T = (A * B)^T
    // This gives us the correct result in column-major format
    cblas_dgemm(CblasColMajor, CblasNoTrans, CblasNoTrans,
                n, n, n, alpha, A, n, B, n, beta, C, n);
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);

    double time_seconds = duration.count() / 1000000.0;
    double gflops = (2.0 * n * n * n) / (time_seconds * 1e9);

    std::cout << "Matrix size: " << n << "x" << n << std::endl;
    std::cout << "Time taken: " << time_seconds << " seconds" << std::endl;
    std::cout << "Performance: " << gflops << " GFLOPS" << std::endl;

    // Clean up
    delete[] A;
    delete[] B;
    delete[] C;

    return 0;
}

File: matrix_mult.py
#!/usr/bin/env python3

import sys
import numpy as np
import time

def main():
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <matrix_size>", file=sys.stderr)
        sys.exit(1)
    
    try:
        n = int(sys.argv[1])
        if n <= 0:
            print("Matrix size must be positive", file=sys.stderr)
            sys.exit(1)
    except ValueError:
        print("Matrix size must be an integer", file=sys.stderr)
        sys.exit(1)

    try:
        cfg = np.show_config('dicts')['Build Dependencies']['blas']
        blas_library = cfg['name'] + ' ' + cfg['version']
    except:
        blas_library = 'unknown'
    print(f'Blas implementation: {blas_library}')

    # Generate random matrices
    print(f"Generating {n}x{n} matrices...", file=sys.stderr)
    np.random.seed(42)  # For reproducible results
    A = np.random.uniform(-1.0, 1.0, (n, n)).astype(np.float64, order='F')
    B = np.random.uniform(-1.0, 1.0, (n, n)).astype(np.float64, order='F')
    C = np.empty((n, n), dtype=np.float64, order='F')

    # Measure time for matrix multiplication
    print(f"Multiplying {n}x{n} matrices...", file=sys.stderr)
    start_time = time.perf_counter()
    np.matmul(A, B, out=C)
    end_time = time.perf_counter()

    # Calculate timing and performance
    elapsed_time = end_time - start_time
    flops = 2.0 * n * n * n
    gflops = flops / (elapsed_time * 1e9)

    # Output in same format as C++ program
    print(f"Matrix size: {n}x{n}")
    print(f"Time taken: {elapsed_time:.6f} seconds")
    print(f"Performance: {gflops:.2f} GFLOPS")

if __name__ == "__main__":
    main()
